content/blog/2017/2017-02-02-scm-api-2-take2.adoc

```diff
@@ -0,0 +1,124 @@
++
+*Resolution:* The changes in SCM API 2.0.2 introduced to resolve JENKINS-41121 provided a path to resolve this issue without causing a rebuild of all PR branches.
+The BitBucket Branch Source plugin was enhanced to provide an implementation of the new SCM API extension point that connects to BitBucket and retrieves the missing information.
+
+link:https://issues.jenkins-ci.org/browse/JENKINS-41124[JENKINS-41124]::
+During initial testing of the Branch API 2.0.0 release an issue was identified with how Organization Folders handled unusual names.
+None of the existing implementations of the `SCMNavigator` API could generate such unusual names due to form validation on GitHub / BitBucket replacing unusual characters with `-` when creating a repository.
++
+It would be irresponsible to rely on external services sanitizing their input data for the correct operation of Organization Folders.
+Consequently, in Branch API 2.0.0 the names were all transformed into URL safe names, with the original URLs still resolving to the original projects so that any existing saved links would remain functional.
++
+Quite a number of people objected to this change of URL scheme.
++
+*Resolution:* There has been a convention in Jenkins that the on-disk storage structure for jobs mirrors the URL structure.
+This is only a convention and there is nothing specific in the code that mandates following the convention.
++
+The Folders Plugin was enhanced to allow for computed folders (where the item names are provided by an external source) to provide a strategy to use when generating the on-disk storage names as well as the URL component names for the folder's child items.
++
+The Branch API plugin was enhanced to use this new strategy for name transformation.
++
+The net effect of this change is that the URLs remain the same as for 1.x but the on-disk storage uses transformed names that are future proofed against any new SCMNavigator implementations where the backing service allows names that are problematic to use as filesystem directory names.
++
+*Side-effect:* The Branch API 2.0.0 approach handled the transformation of names by renaming the items using the Jenkins Item rename API.
+The Branch API 2.0.2 approach does not rename the child items as it is only the on-disk storage location that is moved.
+This means that the Jenkins Item rename API cannot be used.
++
+At this time, the only known side-effect is in the  link:https://plugins.jenkins.io/jobConfigHistory[Job Configuration History] plugin.
+The configuration history of each child item will still be tracked going forward after the upgrade.
+The pre-upgrade configuration history is also retained.
+Because the Jenkins Item rename API cannot be used to flag the configuration file location change, there is no association between the pre-upgrade history chain and the post-upgrade history chain.
+
+The above issues are all resolved by the following plugin releases:
```

Comment:
Can you move this up to line 60?  It doesn't seem like something we want buried.